c/src/space.rs

```diff
@@ -421,6 +527,28 @@ pub extern "C" fn space_subst(space: *const space_t,
     return_atoms(&results, callback, context);
 }
 
+/// NOTE: This function will return -1 for spaces on which it is impossible or impractical
+/// to determine the number of atoms
+#[no_mangle]
+pub extern "C" fn space_atom_count(space: *const space_t) -> isize {
+    match unsafe { (*space).borrow().atom_count() } {
+        Some(count) => count as isize,
+        None => -1
+    }
+}
+
+/// NOTE: This function will do nothing for spaces on which space_atom_count() returns a
+/// non-positive number
+#[no_mangle]
+pub extern "C" fn space_iterate(space: *const space_t,
```

Comment:
What do you propose for the interface?

We could panic, and catch it at a higher level, but I didn't see any code for that in the project so far.  We could provide an error callback in the C API, to be called when an exception like this is encountered.

My feeling is that this should be a "soft error" because the higher levels of the program might not know what kind of space implementation is beneath them, and they shouldn't depend on space_iterate if they don't know what kind of space they're dealing with.  Consider a distributed space with opportunistic concurrency, spread across the whole internet.

Reply:
I am sorry, I was not specific enough. At first my comment was about this method https://github.com/luketpeterson/hyperon-experimental/blob/e30714958283e749375a4fe3bc54aae42c8cd1ca/python/hyperon/base.py#LL33C12-L33C12 which could throw an exception in case the atomspace doesn't have an iterator implementation. But it is a bit pointless in this form because we know `GroundingSpace` has this implementation.

So I am rewording it. Using Rust and `space_api_t` API user can detect the iterator is not implemented by checking the result of the `new_atom_iterator_state`. But this function is not implemented in `hyperonpy.cpp` and thus Python API cannot detect this. What do you think about removing `space_list` and `space_iterate` functions and exposing `new_atom_iterator_state`, `next_atom` and `free_atom_iterator_state` instead? Maybe it is better to do it as a part of another PR which would introduce the Python API for implementing spaces.

Reply:
> space_api_t API user can detect the iterator is not implemented by checking the result of the new_atom_iterator_state. ... 

In the future, we might have Rust-implemented spaces that are not iterable, called from C and Python (and vise-versa if we have a Rust client loading C or Python modules), so we should work to keep the space implementation (space_api_t) separate from the space interface (space_t), and make sure it is possible to do the required checking from space_t.

As you say, Python currently only supports creating a GroundingSpace, so it's currently a non-issue in the code.

I committed 8db6072 with a ToDo, so I (or somebody else) will remember to do the right thing when Python can instantiate other Space implementations.

Reply:
Filed https://github.com/trueagi-io/hyperon-experimental/issues/308

Reply:
> We should work to keep the space implementation (`space_api_t`) separate from the space interface (space_t), and make sure it is possible to do the required checking from `space_t`.

I don't get why you considering `space_t` as an interface and `space_api_t` as an implementation. To me `space_api_t` is an interface which C code should implement in order to have compatible space implementation. But I think we can discuss it further when #308 is being implemented. 

Thanks for a TODO commit, I think it should work.